File: app/src/main/java/name/monwf/customiuizer/utils/AppHelper.java
package name.monwf.customiuizer.utils;

import static name.monwf.customiuizer.utils.Helpers.getAppName;

import android.content.Context;
import android.content.DialogInterface;
import android.content.SharedPreferences;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.text.InputType;
import android.util.Log;
import android.util.Pair;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.appcompat.app.AlertDialog;

import java.util.Locale;
import java.util.Map;
import java.util.Set;

import io.github.libxposed.service.RemotePreferences;
import name.monwf.customiuizer.R;
import name.monwf.customiuizer.mods.GlobalActions;

public class AppHelper {
    public static final String prefsName = "customiuizer_prefs";
    public static SharedPreferences appPrefs = null;
    public static boolean moduleActive = false;

    public static RemotePreferences remotePrefs = null;
    private static final String TAG = "LSPosed-Bridge";
    public static boolean silentSync = false;
    public static String RESTORED_FROM_BACKUP = "restored_from_backup";

    public static void log(String line) {
        Log.i(TAG, "[CustoMIUIzer] " + line);
    }

    public static void log(Throwable t) {
        String logStr = Log.getStackTraceString(t);
        Log.e(TAG, "[CustoMIUIzer] " + logStr);
    }

    public static void log(String mod, String line) {
        Log.i(TAG, "[CustoMIUIzer][" + mod + "] " + line);
    }

    public static void log(String mod, Throwable t) {
        String logStr = Log.getStackTraceString(t);
        Log.e(TAG, "[CustoMIUIzer][" + mod + "] " + logStr);
    }

    public static SharedPreferences getSharedPrefs(Context context, boolean protectedStorage) {
        if (protectedStorage) context = getProtectedContext(context);
        return context.getSharedPreferences(prefsName, Context.MODE_PRIVATE);
    }

    public static int getIntOfAppPrefs(String key, int defValue) {
        if (!key.startsWith("pref_key_")) {
            key = "pref_key_" + key;
        }
        return appPrefs.getInt(key, defValue);
    }

    public static String getStringOfAppPrefs(String key, String defValue) {
        if (!key.startsWith("pref_key_")) {
            key = "pref_key_" + key;
        }
        return appPrefs.getString(key, defValue);
    }

    public static int getStringAsIntOfAppPrefs(String key, int defValue) {
        if (!key.startsWith("pref_key_")) {
            key = "pref_key_" + key;
        }
        String prefValue = getStringOfAppPrefs(key, null);
        if (prefValue == null) return defValue;
        return Integer.parseInt(prefValue);
    }

    public static Set<String> getStringSetOfAppPrefs(String key, Set<String> defValue) {
        if (!key.startsWith("pref_key_")) {
            key = "pref_key_" + key;
        }
        return appPrefs.getStringSet(key, defValue);
    }

    public static boolean getBooleanOfAppPrefs(String key) {
        return getBooleanOfAppPrefs(key, false);
    }
    public static boolean getBooleanOfAppPrefs(String key, boolean defValue) {
        if (!key.startsWith("pref_key_")) {
            key = "pref_key_" + key;
        }
        return appPrefs.getBoolean(key, defValue);
    }

    public static synchronized Context getLocaleContext(Context context) throws Throwable {
        if (appPrefs != null) {
            String locale = getStringOfAppPrefs("pref_key_miuizer_locale", "auto");
            if ("auto".equals(locale) || "1".equals(locale)) return context;
            Configuration config = context.getResources().getConfiguration();
            config.setLocale(Locale.forLanguageTag(locale));
            return context.createConfigurationContext(config);
        } else {
            return context;
        }
    }

    public static synchronized Context getProtectedContext(Context context) {
        return getProtectedContext(context, null);
    }

    public static synchronized Context getProtectedContext(Context context, Configuration config) {
        try {
            Context mContext = context.isDeviceProtectedStorage() ? context : context.createDeviceProtectedStorageContext();
            return getLocaleContext(config == null ? mContext : mContext.createConfigurationContext(config));
        } catch (Throwable t) {
            return context;
        }
    }

    public static void showInputDialog(Context context, final String key, int titleRes, int summRes, int maxLines, Helpers.InputCallback callback) {
        showInputDialog(context, key, titleRes, summRes, maxLines, callback, true);
    }

    public static void showInputDialog(Context context, final String key, int titleRes, int summRes, int maxLines, Helpers.InputCallback callback, boolean prefDefault) {
        AlertDialog.Builder builder = new AlertDialog.Builder(context);
        builder.setTitle(titleRes);
        final EditText input = new EditText(context);
        if (prefDefault) {
            input.setText(getStringOfAppPrefs(key, ""));
        }
        else {
            input.setText(key);
        }

        if (maxLines > 1) {
            input.setSingleLine(false);
            input.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_FLAG_MULTI_LINE);
        }
        else {
            input.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_NORMAL);
        }

        LinearLayout container = new LinearLayout(context);
        int horizPadding = context.getResources().getDimensionPixelSize(R.dimen.preference_item_child_padding);
        container.setPadding(horizPadding, 0, horizPadding, 0);
        container.setOrientation(LinearLayout.VERTICAL);
        if (summRes > 0) {
            final TextView msg = new TextView(context);
            msg.setText(summRes);
            msg.setLayoutParams(new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT));
            container.addView(msg);
        }
        container.addView(input);
        builder.setView(container);
        builder.setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                callback.onInputFinished(key, input.getText().toString());
            }
        });
        builder.setNegativeButton(android.R.string.cancel, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                dialog.cancel();
            }
        });
        builder.show();
    }

    public static void addStringPair(Set<String> hayStack, String needle1, String needle2) {
        if (hayStack != null) hayStack.add(needle1 + "|" + needle2);
    }

    public static void removeStringPair(Set<String> hayStack, String needle) {
        if (hayStack != null)
            for (String pair: hayStack) {
                String[] needles = pair.split("\\|", 2);
                if (needles[0].equals(needle)) {
                    hayStack.remove(pair);
                    return;
                }
            }
    }
    public static Pair<String, String> getActionNameLocal(Context context, String key) {
        try {
            int action = AppHelper.getIntOfAppPrefs(key + "_action", 1);
            Resources modRes = context.getResources();
            Pair<String, String> pair = null;
            int resId = GlobalActions.getActionResId(action);
            if (resId != 0)
                pair = new Pair<>(modRes.getString(resId), "");
            else if (action == 8)
                pair = new Pair<>(modRes.getString(R.string.array_global_actions_launch), (String)getAppName(context, AppHelper.getStringOfAppPrefs(key + "_app", ""), true));
            else if (action == 9)
                pair = new Pair<>(modRes.getString(R.string.array_global_actions_shortcut), AppHelper.getStringOfAppPrefs(key + "_shortcut_name", ""));
            else if (action == 10) {
                int what = AppHelper.getIntOfAppPrefs(key + "_toggle", 0);
                String toggle = modRes.getString(R.string.array_global_actions_toggle);
                switch (what) {
                    case 1: pair = new Pair<>(toggle, modRes.getString(R.string.array_global_toggle_wifi)); break;
                    case 2: pair = new Pair<>(toggle, modRes.getString(R.string.array_global_toggle_bt)); break;
                    case 3: pair = new Pair<>(toggle, modRes.getString(R.string.array_global_toggle_gps)); break;
                    case 4: pair = new Pair<>(toggle, modRes.getString(R.string.array_global_toggle_nfc)); break;
                    case 5: pair = new Pair<>(toggle, modRes.getString(R.string.array_global_toggle_sound)); break;
                    case 6: pair = new Pair<>(toggle, modRes.getString(R.string.array_global_toggle_brightness)); break;
                    case 7: pair = new Pair<>(toggle, modRes.getString(R.string.array_global_toggle_rotation)); break;
                    case 8: pair = new Pair<>(toggle, modRes.getString(R.string.array_global_toggle_torch)); break;
                    case 9: pair = new Pair<>(toggle, modRes.getString(R.string.array_global_toggle_mobiledata)); break;
                }
            } else if (action == 20) {
                String pref = AppHelper.getStringOfAppPrefs(key + "_activity", "");
                String name = (String)getAppName(context, pref);
                if (name == null || name.isEmpty()) name = (String)getAppName(context, pref, true);
                pair = new Pair<>(modRes.getString(R.string.array_global_actions_activity), name);
            }
            return pair;
        } catch (Throwable t) {
            t.printStackTrace();
            return null;
        }
    }
    public static void syncPrefsToAnother(Map<String, ?> entries, SharedPreferences prefs, int clearType, Set<String> ignoreKeys, boolean commitAction) {
        if (entries == null || entries.isEmpty()) return;
        SharedPreferences.Editor prefEdit = prefs.edit();
        if (clearType == 1) {
            prefEdit.clear();
        }
        else if (clearType == 2) {
            for (String key:prefs.getAll().keySet()) {
                if (!entries.containsKey(key)) {
                    prefEdit.remove(key);
                }
            }
        }
        for (Map.Entry<String, ?> entry: entries.entrySet()) {
            String key = entry.getKey();
            if (ignoreKeys != null && ignoreKeys.contains(key)) continue;
            Object val = entry.getValue();

            if (val instanceof Boolean)
                prefEdit.putBoolean(key, (Boolean)val);
            else if (val instanceof Float)
                prefEdit.putFloat(key, (Float)val);
            else if (val instanceof Integer)
                prefEdit.putInt(key, (Integer)val);
            else if (val instanceof Long)
                prefEdit.putLong(key, (Long)val);
            else if (val instanceof String)
                prefEdit.putString(key, ((String)val));
            else if (val instanceof Set<?>)
                prefEdit.putStringSet(key, ((Set<String>)val));
        }
        if (commitAction) {
            prefEdit.commit();
        }
        else {
            prefEdit.apply();
        }
    }
}

File: app/src/main/java/tv/withaibuild/customiuizer/utils/AppHelper.kt
package tv.withaibuild.customiuizer.utils

import android.content.Context
import android.content.SharedPreferences
import android.content.res.Configuration
import android.content.res.Resources
import android.text.InputType
import android.util.Log
import android.util.Pair
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import io.github.libxposed.service.RemotePreferences
import tv.withaibuild.customiuizer.R
import tv.withaibuild.customiuizer.mods.GlobalActions
import java.util.Locale

object AppHelper {
    const val prefsName = "customiuizer_prefs"
    private const val TAG = "LSPosed-Bridge"

    @JvmField
    var appPrefs: SharedPreferences? = null

    @JvmField
    var moduleActive = false

    @JvmField
    var remotePrefs: RemotePreferences? = null

    @JvmField
    var silentSync = false

    @JvmField
    var installedAppsList: ArrayList<AppData> = ArrayList()

    enum class SettingsType {
        Preference, Edit
    }

    enum class AppAdapterType {
        Default, Standalone, Mutli, CustomTitles, Activities
    }

    enum class ActionBarType {
        HomeUp, Edit
    }

    @JvmStatic
    fun log(line: String) {
        Log.i(TAG, "[Pengeek] $line")
    }

    @JvmStatic
    fun log(t: Throwable) {
        Log.e(TAG, "[Pengeek] " + Log.getStackTraceString(t))
    }

    @JvmStatic
    fun log(mod: String, line: String) {
        Log.i(TAG, "[Pengeek][$mod] $line")
    }

    @JvmStatic
    fun log(mod: String, t: Throwable) {
        Log.e(TAG, "[Pengeek][$mod] " + Log.getStackTraceString(t))
    }

    @JvmStatic
    fun getSharedPrefs(context: Context, protectedStorage: Boolean): SharedPreferences {
        val ctx = if (protectedStorage) getProtectedContext(context) else context
        return ctx.getSharedPreferences(prefsName, Context.MODE_PRIVATE)
    }

    @JvmStatic
    fun getIntOfAppPrefs(key: String?, defValue: Int): Int {
        if (key == null) return defValue
        return appPrefs?.getInt(prefixKey(key), defValue) ?: defValue
    }

    @JvmStatic
    fun getStringOfAppPrefs(key: String?, defValue: String?): String? {
        if (key == null) return defValue
        return appPrefs?.getString(prefixKey(key), defValue) ?: defValue
    }

    @JvmStatic
    fun getStringAsIntOfAppPrefs(key: String?, defValue: Int): Int {
        val prefValue = getStringOfAppPrefs(key, null)
        return if (prefValue == null) defValue else Integer.parseInt(prefValue)
    }

    @JvmStatic
    fun getStringSetOfAppPrefs(key: String?, defValue: Set<String>): MutableSet<String> {
        if (key == null) return defValue.toMutableSet()
        val set = appPrefs?.getStringSet(prefixKey(key), defValue) ?: defValue
        return set.toMutableSet()
    }

    @JvmStatic
    @JvmOverloads
    fun getBooleanOfAppPrefs(key: String?, defValue: Boolean = false): Boolean {
        if (key == null) return defValue
        return appPrefs?.getBoolean(prefixKey(key), defValue) ?: defValue
    }

    @JvmStatic
    @Synchronized
    fun getLocaleContext(context: Context): Context {
        val prefs = appPrefs ?: return context
        val locale = prefs.getString("pref_key_miuizer_locale", "auto") ?: "auto"
        if (locale == "auto" || locale == "1") return context
        val config = Configuration(context.resources.configuration)
        config.setLocale(Locale.forLanguageTag(locale))
        return context.createConfigurationContext(config)
    }

    @JvmStatic
    @JvmOverloads
    @Synchronized
    fun getProtectedContext(context: Context, config: Configuration? = null): Context {
        return try {
            val mContext = if (context.isDeviceProtectedStorage) context else context.createDeviceProtectedStorageContext()
            getLocaleContext(if (config == null) mContext else mContext.createConfigurationContext(config))
        } catch (t: Throwable) {
            context
        }
    }

    @JvmStatic
    fun showInputDialog(
        context: Context?,
        key: String?,
        titleRes: Int,
        summRes: Int,
        maxLines: Int,
        callback: Helpers.InputCallback
    ) {
        showInputDialog(context, key, titleRes, summRes, maxLines, callback, true)
    }

    @JvmStatic
    fun showInputDialog(
        context: Context?,
        key: String?,
        titleRes: Int,
        summRes: Int,
        maxLines: Int,
        callback: Helpers.InputCallback,
        prefDefault: Boolean
    ) {
        if (context == null || key == null) return
        val builder = AlertDialog.Builder(context)
        builder.setTitle(titleRes)
        val input = EditText(context)
        input.setText(if (prefDefault) getStringOfAppPrefs(key, "") else key)

        if (maxLines > 1) {
            input.isSingleLine = false
            input.inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_FLAG_MULTI_LINE
        } else {
            input.inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_NORMAL
        }

        val container = LinearLayout(context)
        val horizPadding = context.resources.getDimensionPixelSize(R.dimen.preference_item_child_padding)
        container.setPadding(horizPadding, 0, horizPadding, 0)
        container.orientation = LinearLayout.VERTICAL
        if (summRes > 0) {
            val msg = TextView(context)
            msg.setText(summRes)
            msg.layoutParams = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT
            )
            container.addView(msg)
        }
        container.addView(input)
        builder.setView(container)
        builder.setPositiveButton(android.R.string.ok) { _, _ ->
            callback.onInputFinished(key, input.text.toString())
        }
        builder.setNegativeButton(android.R.string.cancel) { dialog, _ ->
            dialog.cancel()
        }
        builder.show()
    }

    @JvmStatic
    fun addStringPair(hayStack: MutableSet<String>?, needle1: String, needle2: String) {
        hayStack?.add("$needle1|$needle2")
    }

    @JvmStatic
    fun removeStringPair(hayStack: MutableSet<String>?, needle: String) {
        hayStack ?: return
        val it = hayStack.iterator()
        while (it.hasNext()) {
            val pair = it.next()
            if (pair.substringBefore("|") == needle) {
                it.remove()
                return
            }
        }
    }

    @JvmStatic
    fun getActionNameLocal(context: Context, key: String): Pair<String, String>? {
        return try {
            val action = getIntOfAppPrefs(key + "_action", 1)
            val modRes = context.resources
            var pair: Pair<String, String>? = null
            val resId = GlobalActions.getActionResId(action)
            if (resId != 0) {
                pair = Pair(modRes.getString(resId), "")
            } else if (action == 8) {
                pair = Pair(
                    modRes.getString(R.string.array_global_actions_launch),
                    Helpers.getAppName(context, getStringOfAppPrefs(key + "_app", "") ?: "", true)?.toString() ?: ""
                )
            } else if (action == 9) {
                pair = Pair(
                    modRes.getString(R.string.array_global_actions_shortcut),
                    getStringOfAppPrefs(key + "_shortcut_name", "") ?: ""
                )
            } else if (action == 10) {
                val what = getIntOfAppPrefs(key + "_toggle", 0)
                val toggle = modRes.getString(R.string.array_global_actions_toggle)
                pair = when (what) {
                    1 -> Pair(toggle, modRes.getString(R.string.array_global_toggle_wifi))
                    2 -> Pair(toggle, modRes.getString(R.string.array_global_toggle_bt))
                    3 -> Pair(toggle, modRes.getString(R.string.array_global_toggle_gps))
                    4 -> Pair(toggle, modRes.getString(R.string.array_global_toggle_nfc))
                    5 -> Pair(toggle, modRes.getString(R.string.array_global_toggle_sound))
                    6 -> Pair(toggle, modRes.getString(R.string.array_global_toggle_brightness))
                    7 -> Pair(toggle, modRes.getString(R.string.array_global_toggle_rotation))
                    8 -> Pair(toggle, modRes.getString(R.string.array_global_toggle_torch))
                    9 -> Pair(toggle, modRes.getString(R.string.array_global_toggle_mobiledata))
                    else -> null
                }
            } else if (action == 20) {
                val pref = getStringOfAppPrefs(key + "_activity", "") ?: ""
                var name = Helpers.getAppName(context, pref)?.toString()
                if (name.isNullOrEmpty()) name = Helpers.getAppName(context, pref, true)?.toString()
                pair = Pair(modRes.getString(R.string.array_global_actions_activity), name ?: "")
            }
            pair
        } catch (t: Throwable) {
            t.printStackTrace()
            null
        }
    }

    @JvmStatic
    fun syncPrefsToAnother(
        entries: Map<String, Any?>?,
        prefs: SharedPreferences,
        clearType: Int,
        ignoreKeys: Set<String>?,
        commitAction: Boolean
    ) {
        if (entries.isNullOrEmpty()) return
        val prefEdit = prefs.edit()
        when (clearType) {
            1 -> prefEdit.clear()
            2 -> {
                for (k in prefs.all.keys) {
                    if (!entries.containsKey(k)) {
                        prefEdit.remove(k)
                    }
                }
            }
        }
        for ((key, value) in entries) {
            if (ignoreKeys != null && ignoreKeys.contains(key)) continue
            if (value == null) continue
            when (value) {
                is Boolean -> prefEdit.putBoolean(key, value)
                is Float -> prefEdit.putFloat(key, value)
                is Int -> prefEdit.putInt(key, value)
                is Long -> prefEdit.putLong(key, value)
                is String -> prefEdit.putString(key, value)
                is Set<*> -> @Suppress("UNCHECKED_CAST") prefEdit.putStringSet(key, value as Set<String>)
            }
        }
        if (commitAction) {
            prefEdit.commit()
        } else {
            prefEdit.apply()
        }
    }

    private fun prefixKey(key: String): String {
        return if (key.startsWith("pref_key_")) key else "pref_key_$key"
    }
}
